File: gui/src/components/Commissioning/pathLossApplication.ts
export type PathLossApplicationStatus = 'applied' | 'not_applied' | 'unknown'
export type PathLossApplicationProvenance = 'real' | 'simulated' | 'unknown' | 'missing'
export type PathLossApplicationReason =
  | 'selected'
  | 'rejected_untrusted'
  | 'missing'
  | 'expired'
  | 'frequency_mismatch'
  | 'operating_mode_mismatch'
  | 'legacy_unclassified'

export interface PathLossApplication {
  schema_version: 1
  status: PathLossApplicationStatus
  provenance: PathLossApplicationProvenance
  reason: PathLossApplicationReason
  gate_mode: 'strict' | 'operator_bypass' | 'mock_not_applicable'
  certificate_id: string | null
  value_disclosure: 'verified' | 'hidden_unverified' | 'none'
}

export interface PathLossApplicationView {
  title: string
  message: string
  color: 'green' | 'yellow'
  certificateId: string | null
  sourceLabel: string
  showCompensationValue: boolean
}

const fields = [
  'certificate_id',
  'gate_mode',
  'provenance',
  'reason',
  'schema_version',
  'status',
  'value_disclosure',
]

const legacyUnknown = (): PathLossApplication => ({
  schema_version: 1,
  status: 'unknown',
  provenance: 'unknown',
  reason: 'legacy_unclassified',
  gate_mode: 'strict',
  certificate_id: null,
  value_disclosure: 'none',
})

const isRecord = (value: unknown): value is Record<string, unknown> => (
  typeof value === 'object' && value !== null && !Array.isArray(value)
)

export function parsePathLossApplication(value: unknown): PathLossApplication {
  if (!isRecord(value) || Object.keys(value).sort().join('|') !== fields.join('|')) {
    return legacyUnknown()
  }
  if (
    value.schema_version !== 1
    || !['strict', 'operator_bypass', 'mock_not_applicable'].includes(String(value.gate_mode))
  ) {
    return legacyUnknown()
  }

  const certificateId = value.certificate_id
  const hasCertificateId = typeof certificateId === 'string' && certificateId.length > 0
  if (value.status === 'applied') {
    if (
      !['real', 'simulated', 'unknown'].includes(String(value.provenance))
      || value.reason !== 'selected'
      || !hasCertificateId
    ) {
      return legacyUnknown()
    }
    const expectedDisclosure = value.provenance === 'real' ? 'verified' : 'hidden_unverified'
    if (value.value_disclosure !== expectedDisclosure) return legacyUnknown()
    return value as unknown as PathLossApplication
  }

  if (value.status === 'not_applied' && value.reason === 'rejected_untrusted') {
    if (
      !['simulated', 'unknown'].includes(String(value.provenance))
      || !hasCertificateId
      || value.value_disclosure !== 'none'
    ) {
      return legacyUnknown()
    }
    return value as unknown as PathLossApplication
  }

  if (
    value.status === 'not_applied'
    && ['missing', 'expired', 'frequency_mismatch', 'operating_mode_mismatch'].includes(String(value.reason))
    && value.provenance === 'missing'
    && value.certificate_id === null
    && value.value_disclosure === 'none'
  ) {
    return value as unknown as PathLossApplication
  }

  return legacyUnknown()
}

export function describePathLossApplication(value: unknown): PathLossApplicationView {
  const application = parsePathLossApplication(value)
  const showCompensationValue = (
    application.status === 'applied'
    && application.provenance === 'real'
    && application.value_disclosure === 'verified'
  )
  const sourceLabels: Record<PathLossApplicationProvenance, string> = {
    real: '真实来源',
    simulated: '模拟来源',
    unknown: '来源未知',
    missing: '无匹配证书',
  }
  let title = '路损应用状态未知'
  let message = '历史记录无法证明是否应用路损补偿；补偿数值不展示。'

  if (showCompensationValue) {
    title = '路损补偿已验证'
    message = '已应用经验证的路损补偿。'
  } else if (application.status === 'applied' && application.provenance === 'simulated') {
    title = '已应用模拟路损证书'
    message = '已应用模拟路损证书用于流程演练；数值不进入正式结果。'
  } else if (application.status === 'applied' && application.provenance === 'unknown') {
    title = '路损补偿已应用，但来源未知'
    message = '已应用路损补偿；证书来源未知，补偿数值不展示，结果不参与正式判定。'
  } else if (application.reason === 'rejected_untrusted') {
    title = '路损证书未应用'
    message = '检测到路损证书，但因来源未验证未应用；本次结果未补偿。'
  } else if (application.reason === 'expired') {
    title = '路损证书已过期'
    message = '匹配的路损证书已过期；本次结果未补偿。'
  } else if (application.reason === 'frequency_mismatch') {
    title = '路损证书频率不匹配'
    message = '现有证书与本次频率不匹配；本次结果未补偿。'
  } else if (application.reason === 'operating_mode_mismatch') {
    title = '路损证书 RF 模式不匹配'
    message = '现有证书与本次 RF operating mode 不匹配；本次结果未补偿。'
  } else if (application.reason === 'missing') {
    title = '未找到路损证书'
    message = '未找到匹配的路损证书；本次结果未补偿。'
  }

  return {
    title,
    message,
    color: showCompensationValue ? 'green' : 'yellow',
    certificateId: application.certificate_id,
    sourceLabel: sourceLabels[application.provenance],
    showCompensationValue,
  }
}

export function describePathLossSelection(reason: unknown, valid: boolean): string {
  if (valid && reason === 'selected') return '有效（已选中）'
  const labels: Record<string, string> = {
    missing: '未找到匹配证书',
    expired: '匹配证书已过期',
    frequency_mismatch: '证书频率不匹配',
    operating_mode_mismatch: '证书 RF 模式不匹配',
    selected: '已选中但未通过来源门',
  }
  return labels[String(reason)] ?? '状态未知'
}

File: api-service/app/services/mimo_ota/path_loss_application.py
"""路损补偿的应用事实快照。

应用、证书来源与正式可信门是三个独立事实。历史记录缺失或畸形时只能降级为
``unknown``，不得从旧补偿数值或证书 ID 反推。
"""
from __future__ import annotations

from typing import Any, Mapping, Optional


SCHEMA_VERSION = 1
_GATE_MODES = {"strict", "operator_bypass", "mock_not_applicable"}
_MISSING_REASONS = {
    "missing",
    "expired",
    "frequency_mismatch",
    "operating_mode_mismatch",
}
_FIELDS = {
    "schema_version",
    "status",
    "provenance",
    "reason",
    "gate_mode",
    "certificate_id",
    "value_disclosure",
}


def _legacy_unknown() -> dict[str, Any]:
    return {
        "schema_version": SCHEMA_VERSION,
        "status": "unknown",
        "provenance": "unknown",
        "reason": "legacy_unclassified",
        "gate_mode": "strict",
        "certificate_id": None,
        "value_disclosure": "none",
    }


def _certificate_provenance(certificate: Any) -> str:
    use_mock = getattr(certificate, "use_mock", None)
    if use_mock is False:
        return "real"
    if use_mock is True:
        return "simulated"
    return "unknown"


def build_path_loss_application(
    *,
    selected_certificate: Optional[Any],
    applied_certificate: Optional[Any],
    selection_reason: str,
    gate_mode: str,
) -> dict[str, Any]:
    """从当次选择与实际应用对象构造不可变语义快照。"""
    if gate_mode not in _GATE_MODES:
        raise ValueError(f"Unknown path-loss gate mode: {gate_mode}")

    if applied_certificate is not None:
        if selected_certificate is None or selection_reason != "selected":
            raise ValueError("An applied certificate must be the selected certificate")
        if getattr(applied_certificate, "id", None) != getattr(
            selected_certificate, "id", None
        ):
            raise ValueError("Applied and selected path-loss certificates differ")
        provenance = _certificate_provenance(applied_certificate)
        return {
            "schema_version": SCHEMA_VERSION,
            "status": "applied",
            "provenance": provenance,
            "reason": "selected",
            "gate_mode": gate_mode,
            "certificate_id": str(applied_certificate.id),
            "value_disclosure": (
                "verified" if provenance == "real" else "hidden_unverified"
            ),
        }

    if selected_certificate is not None:
        if selection_reason != "selected":
            raise ValueError("A selected certificate requires selection reason 'selected'")
        return {
            "schema_version": SCHEMA_VERSION,
            "status": "not_applied",
            "provenance": _certificate_provenance(selected_certificate),
            "reason": "rejected_untrusted",
            "gate_mode": gate_mode,
            "certificate_id": str(selected_certificate.id),
            "value_disclosure": "none",
        }

    if selection_reason not in _MISSING_REASONS:
        raise ValueError(f"Unknown path-loss selection reason: {selection_reason}")
    return {
        "schema_version": SCHEMA_VERSION,
        "status": "not_applied",
        "provenance": "missing",
        "reason": selection_reason,
        "gate_mode": gate_mode,
        "certificate_id": None,
        "value_disclosure": "none",
    }


def parse_path_loss_application(value: Any) -> dict[str, Any]:
    """验证持久化快照；任何未知/非法组合都保守降级。"""
    if not isinstance(value, Mapping) or set(value) != _FIELDS:
        return _legacy_unknown()
    parsed = dict(value)
    if (
        type(parsed.get("schema_version")) is not int
        or parsed["schema_version"] != SCHEMA_VERSION
    ):
        return _legacy_unknown()
    if parsed.get("gate_mode") not in _GATE_MODES:
        return _legacy_unknown()

    status = parsed.get("status")
    provenance = parsed.get("provenance")
    reason = parsed.get("reason")
    certificate_id = parsed.get("certificate_id")
    disclosure = parsed.get("value_disclosure")

    if status == "applied":
        if (
            provenance not in {"real", "simulated", "unknown"}
            or reason != "selected"
            or not isinstance(certificate_id, str)
            or not certificate_id
        ):
            return _legacy_unknown()
        if (
            provenance in {"simulated", "unknown"}
            and parsed["gate_mode"] != "mock_not_applicable"
        ):
            return _legacy_unknown()
        expected_disclosure = "verified" if provenance == "real" else "hidden_unverified"
        return parsed if disclosure == expected_disclosure else _legacy_unknown()

    if status == "not_applied" and reason == "rejected_untrusted":
        if (
            provenance not in {"simulated", "unknown"}
            or not isinstance(certificate_id, str)
            or not certificate_id
            or disclosure != "none"
            or parsed["gate_mode"] == "mock_not_applicable"
        ):
            return _legacy_unknown()
        return parsed

    if status == "not_applied" and reason in _MISSING_REASONS:
        if provenance == "missing" and certificate_id is None and disclosure == "none":
            return parsed

    if parsed == _legacy_unknown():
        return parsed
    return _legacy_unknown()


def path_loss_application_is_formally_verified(value: Any) -> bool:
    """正式消费者统一白名单：必须是可解析的 applied explicit-real 快照。"""
    application = parse_path_loss_application(value)
    return (
        application["status"] == "applied"
        and application["provenance"] == "real"
        and application["reason"] == "selected"
        and application["value_disclosure"] == "verified"
    )


def path_loss_application_message(value: Any) -> str:
    """生成人读叙事；故意不接收或格式化任何补偿数值。"""
    application = parse_path_loss_application(value)
    status = application["status"]
    provenance = application["provenance"]
    reason = application["reason"]

    if status == "applied" and provenance == "real":
        return "已应用经验证的路损补偿。"
    if status == "applied" and provenance == "simulated":
        return "已应用模拟路损证书用于流程演练；数值不进入正式结果。"
    if status == "applied" and provenance == "unknown":
        return "已应用路损补偿；证书来源未知，补偿数值不展示，结果不参与正式判定。"
    if reason == "rejected_untrusted":
        return "检测到路损证书，但因来源未验证未应用；本次结果未补偿。"
    if reason == "expired":
        return "匹配的路损证书已过期；本次结果未补偿。"
    if reason == "frequency_mismatch":
        return "现有证书与本次频率不匹配；本次结果未补偿。"
    if reason == "operating_mode_mismatch":
        return "现有证书与本次 RF operating mode 不匹配；本次结果未补偿。"
    if reason == "missing":
        return "未找到匹配的路损证书；本次结果未补偿。"
    return "历史记录无法证明是否应用路损补偿；补偿数值不展示。"
